### DISCUSSION TITLE
how to train controlnet with lora?

### DISCUSSION THREAD

**Discussion by akk-123** [huggingface/diffusers#7107] [General]
train full controlnet need much resource and time, so how to train controlnet with lora?

**Comment by sayakpaul** [comment]
You can adapt our ControlNet training script and refer to the LoRA training scripts to do this. 

For example, after initializing he `ControlNetModel`, you could add the following LoRA config:

```python
from peft import LoraConfig

lora_config = LoraConfig(
    r=args.rank,
    lora_alpha=args.rank,
    init_lora_weights="gaussian",
    target_modules=["to_k", "to_q", "to_v", "to_out.0"],
)
controlnet.add_adapter(lora_config)
```

Does that help?

**Comment by Kinyugo** [comment]
Hello @sayakpaul, I am trying to do a similar thing. I have managed to add the adapters with `add_adapter`. However, I encounter problems when loading the model. According to this guide (https://huggingface.co/docs/transformers/main/en/peft), I should be able to use `save_pretrained` and `from_pretrained`. However, the model is saved without the lora config. So when I call `from_pretrained` I get errors about mismatched weights.

**Reply by sayakpaul** [reply]
The SDXL LoRA script already shows how you need to save and load.

**Comment by lavinal712** [comment]
I have tried to train a controlnet with lora, but after I merge the controlnet model and lora, the weight of controlnet seems not changed, and the output is sd, not controlnet

**Reply by sayakpaul** [reply]
Can you open an issue with that?

**Comment by venkateshtata** [comment]
@sayakpaul I see that the SDXL Lora Script here [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py](https://github.com/huggingface/diffusers/discussions/url), it is plugging lora to the UNet of StableDiffusion, how to add the adapter to the ControlNet's down and mid blocks instead ? Can you refer me to any implementation ?

When I do controlnet.add_adapter I get `AttributeError: 'ControlNetModel' object has no attribute 'add_adapter'` error while passing it to StableDiffusionControlNetPipeline.

**Reply by lavinal712** [reply]
I manage to add lora in ControlNet in SD v1.5. Here is the [repo](https://github.com/lavinal712/control-lora-v3). There are some repo for reference: [control-lora-v2](https://github.com/HighCWu/control-lora-v2), [ControlLoRA](https://github.com/HighCWu/ControlLoRA), and model lib for webui [ControlNet-modules-safetensors](https://huggingface.co/webui/ControlNet-modules-safetensors), [control-lora](https://huggingface.co/stabilityai/control-lora)
Stabilityai didn't release the code of training control-lora, so the repo provide their own managment. If you want to use control-lora in webui, feel free to use the model libs above.

**Comment by sayakpaul** [comment]
Thanks for your requests and discussions. There are a couple of ways doing Control-LoRA training right now with the library. 

## Option 1

Use `peft` native methods. You could define a `LoraConfig` and do something like `get_peft_model(controlnet, config)`. Example is: https://github.com/huggingface/peft/blob/076561bbd32ce8cd649ba4ec05cbc592cf2dd1c6/examples/lora_dreambooth/train_dreambooth.py#L723. 

Here, you would initialize `controlnet` from a corresponding UNet like we do here: https://github.com/huggingface/diffusers/blob/f96e4a16adb4c31bab4c0a3d0d145ed2b086ecb0/examples/controlnet/train_controlnet.py#L818. 

Rest of the training process should be similar. 

To export and use the trained LoRA parameters, you can refer to this notebook:
https://github.com/huggingface/peft/blob/main/examples/lora_dreambooth/lora_dreambooth_inference.ipynb

## Option 2

Have the `ControlNet` inherit from `PeftAdapterMixin` (like [this](https://github.com/huggingface/diffusers/blob/f96e4a16adb4c31bab4c0a3d0d145ed2b086ecb0/src/diffusers/models/unets/unet_2d_condition.py#L71C83-L71C99)). Now, you have access to methods like `add_adapter()` (used [here](https://github.com/huggingface/diffusers/blob/f96e4a16adb4c31bab4c0a3d0d145ed2b086ecb0/examples/dreambooth/train_dreambooth_lora_sdxl.py#L1193)). 

And then you can write your `load_lora_into_controlnet()` method (like [this](https://github.com/huggingface/diffusers/blob/f96e4a16adb4c31bab4c0a3d0d145ed2b086ecb0/src/diffusers/loaders/lora.py#L371)) so that the trained LoRA parameters load properly. 

Hopefully, it makes sense now.

**Reply by venkateshtata** [reply]
Option 2 worked for me, Thank You! I could'nt use option 1 as I had to load a pre-trained Canny Edge ControlNet model.

**Reply by sayakpaul** [reply]
That should work too. So, you would do `ControlNet.from_pretrained()` (passing the path/id to the pretrained model) to initialize `controlnet`.
